xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-filters/xwiki-platform-notifications-filters-default/src/main/java/org/xwiki/notifications/filters/migration/R160000000XWIKI17243DataMigration.java

```diff
@@ -0,0 +1,198 @@
+    private NotificationFilterPreferenceStore notificationFilterPreferenceStore;
+
+    @Inject
+    private DocumentReferenceResolver<String> documentReferenceResolver;
+
+    @Inject
+    private Logger logger;
+
+    @Override
+    public String getDescription()
+    {
+        return "Migrate old WatchListClass xobjects to proper filters.";
+    }
+
+    @Override
+    public XWikiDBVersion getVersion()
+    {
+        return new XWikiDBVersion(160000000);
+    }
+
+    @Override
+    protected void hibernateMigrate() throws DataMigrationException, XWikiException
+    {
+        String statement = "from doc.object(XWiki.WatchListClass) as user";
+        int offset = 0;
+        List<String> results = null;
+        do {
+            try {
+                results = this.queryManager.createQuery(statement, Query.XWQL)
+                    .addFilter(uniqueFilter)
+                    .setOffset(offset)
+                    .setLimit(BATCH_SIZE).execute();
+
+                this.logger.info("Found [{}] users with WatchListClass objects... migrating them.", results.size());
+                for (String result : results) {
+                    DocumentReference userDocReference = this.documentReferenceResolver.resolve(result);
+                    this.migrateUser(userDocReference);
+                }
+            } catch (QueryException e) {
+                throw new DataMigrationException("Error while performing query to find users with WatchList objects",
+                    e);
+            }
+        } while (results != null && results.size() == BATCH_SIZE);
+    }
+
+    private void migrateUser(DocumentReference userDocReference) throws DataMigrationException, XWikiException
+    {
+        XWikiContext context = getXWikiContext();
+        XWikiDocument document = context.getWiki().getDocument(userDocReference, context);
+        EntityReference classReference = CLASS_REFERENCE.appendParent(userDocReference.getWikiReference());
+        List<BaseObject> objects = document.getXObjects(classReference);
+
+        Set<NotificationFilterPreference> results = new HashSet<>();
+        for (BaseObject obj : objects) {
+            if (obj == null) {
+                continue;
+            }
+            getValues(obj, FIELD_WIKIS, NotificationFilterProperty.WIKI, results);
+            getValues(obj, FIELD_SPACES, NotificationFilterProperty.SPACE, results);
+            getValues(obj, FIELD_DOCUMENTS, NotificationFilterProperty.PAGE, results);
+        }
+
+        try {
+            this.notificationFilterPreferenceStore.saveFilterPreferences(userDocReference, results);
+        } catch (NotificationException e) {
+            throw new DataMigrationException(String.format("Error while trying to save [%s] filter preferences",
+                results.size()), e);
+        }
+        if (document.removeXObjects(classReference)) {
+            context.getWiki().saveDocument(document, "Migration of watchlist preferences", context);
```

Comment:
Modifying and saving document during a migration have a tendency to cause problem with the listener which listen to those events because XWiki is not fully initialized. Generally, the migration in such a case is only used to register things to do after the XWiki init is finished.

Reply:
Ok I guess in this case I can use a dedicated TaskConsumer for cleaning up the docs afterwards.